src/exception/exception-api.ts

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2022 The Ontario Institute for Cancer Research. All rights reserved
+ *
+ * This program and the accompanying materials are made available under the terms of
+ * the GNU Affero General Public License v3.0. You should have received a copy of the
+ * GNU Affero General Public License along with this program.
+ *  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+ * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import { Request, Response } from 'express';
+import { HasFullWriteAccess } from '../decorators';
+import { loggerFor } from '../logger';
+import { ControllerUtils, TsvUtils } from '../utils';
+import * as exceptionService from './exception-service';
+
+const L = loggerFor(__filename);
+
+enum ProgramExceptionErrorMessage {
+  TSV_PARSING_FAILED = `This file is formatted incorrectly`,
+}
+
+class ExceptionController {
+  @HasFullWriteAccess()
+  async createProgramException(req: Request, res: Response) {
+    if (!isValidCreateBody(req, res)) {
+      return false;
+    }
+
+    const programId = req.params.programId;
+    const file = req.file;
+
+    let records: ReadonlyArray<TsvUtils.TsvRecordAsJsonObj>;
+
+    try {
+      records = await TsvUtils.tsvToJson(file.path);
+      if (records.length === 0) {
+        throw new Error('TSV has no records!');
+      }
+    } catch (err) {
+      L.error(`Program Exception TSV_PARSING_FAILED`, err);
+      return ControllerUtils.unableToProcess(res, ProgramExceptionErrorMessage.TSV_PARSING_FAILED);
+    }
+
+    const result = await exceptionService.operations.createProgramException({
+      programId,
+      records,
+    });
+
+    if (!result.successful) {
+      return res.status(422).send(result);
+    }
+    return res.status(201).send(result);
```

Comment:
A couple things that need fixing in this block.

1. All the code after the try/catch won't run if we throw an error, so we're doing this extra manipulation of a mutable variable `records` for no reason. Just move all this code into the try block.
2. Won't compile, there's a type error. Specificly, `createProgramException` is expecting `records` to be of type `ProgramExceptionRecord` and we only know that they are `ReadonlyArray<TsvUtils.TsvRecordAsJsonObj>`. So we need to do some validation to make sure our records have the correct structure. This actually turns out to be just tricky enough that I spent 30 minutes writing out a solution (see next comment).

Reply:
First, we need some type validaters with user defined type guards: https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

For ProgramExceptionRecord this might look like:
```TypeScript
export const isProgramExceptionRecord = (input: any): input is ProgramExceptionRecord => {
  return (
    // input must not be null and be an object (typeof null = 'object', amusingly)
    typeof input === 'object' &&
    input !== null &&
    // program_name must exist and be string
    'program_name' in input &&
    typeof input.program_name === 'string' &&
    // schema must exist and be string
    'schema' in input &&
    typeof input.schema === 'string' &&
    // requested_core_field must exist and be string
    'requested_core_field' in input &&
    typeof input.requested_core_field === 'string' &&
    // requested_exception_value must exist and be string and be in enum list
    'requested_exception_value' in input &&
    typeof input.requested_core_field === 'string' &&
    Object.values(ExceptionValue).includes(input.requested_core_field)
  );
};
```

This can validate a single item of any type matches our interface, and tell TypeScript that it is safe to treat it as such.

This doesn't work for the whole array though. In order to type check the whole array we need another tool:

```TypeScript
export const isArrayOf = <T>(input: any[], validator: (_: any) => _ is T): input is T[] => {
  return input.every(validator);
};
export const isReadonlyArrayOf = <T>(input: ReadonlyArray<any>, validator: (_: any) => _ is T): input is ReadonlyArray<T> => {
  return input.every(validator);
};
```

One version for a standard array, and one that can handle the ReadonlyArray type (couldn't combine nicely :( )

Reply:
Finally, we can use these in the `createProgramException` request handler:
```TypeScript
try {
      const records = await TsvUtils.tsvToJson(file.path);
      if (records.length === 0) {
        throw new Error('TSV has no records!');
      }
      if (!isReadonlyArrayOf(records, isProgramExceptionRecord)) {
        throw new Error('TSV is incorrectly structured');
      }
      const result = await exceptionService.operations.createProgramException({
        programId,
        records,
      });

      if (!result.successful) {
        return res.status(422).send(result);
      }
      return res.status(201).send(result);
    } catch (err) {
      L.error(`Program Exception TSV_PARSING_FAILED`, err);
      return ControllerUtils.unableToProcess(res, ProgramExceptionErrorMessage.TSV_PARSING_FAILED);
    }
  }
  ```

Reply:
You'll also need to update the `records` argument type in a couple of your service methods to be `ReadonlyArray<ProgramExceptionRecord>` instead of `ProgramExceptionRecord[]`

Reply:
Very cool. thanks for typing this up. this then does the job of most of the validation in exception-service?

Reply:
Maybe make a new file just for the validations, keep it separate from the application logic.